Objects/perf_trampoline.c

```diff
@@ -0,0 +1,219 @@
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+
+    assert(code_arena->size_left <= code_arena->size);
+    return code_arena_new_code(code_arena);
+}
+
+static inline FILE *
+perf_map_get_file(void)
+{
+    if (perf_map_file) {
+        return perf_map_file;
+    }
+    char filename[100];
+    pid_t pid = getpid();
+    // TODO: %d is incorrect if pid_t is long long
+    snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", pid);
+    perf_map_file = fopen(filename, "a");
+    if (!perf_map_file) {
+        _Py_FatalErrorFormat(__func__, "Couldn't open %s: errno(%d)", filename, errno);
+        return NULL;
+    }
+    return perf_map_file;
+}
+
+static inline int
+perf_map_close(FILE *fp)
+{
+    if (fp) {
+        return fclose(fp);
+    }
+    return 0;
+}
+
+static void
+perf_map_write_entry(FILE *method_file, const void *code_addr,
+                     unsigned int code_size, const char *entry,
+                     const char *file)
+{
+    fprintf(method_file, "%lx %x py::%s:%s\n", (unsigned long)code_addr,
+            code_size, entry, file);
+    fflush(method_file);
+}
+
+static PyObject *
+py_trampoline_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame,
+                        int throw)
+{
+    PyCodeObject *co = frame->f_code;
+    py_trampoline f = NULL;
+    _PyCode_GetExtra((PyObject *)co, extra_code_index, (void **)&f);
+    if (f == NULL) {
+        if (extra_code_index == -1) {
+            extra_code_index = _PyEval_RequestCodeExtraIndex(NULL);
+        }
+        py_trampoline new_trampoline = compile_trampoline();
+        if (new_trampoline == NULL) {
+            return NULL;
+        }
+        FILE *pfile = perf_map_get_file();
+        if (pfile == NULL) {
+            return NULL;
+        }
+        perf_map_write_entry(pfile, new_trampoline, code_arena->code_size,
+                             PyUnicode_AsUTF8(co->co_qualname),
+                             PyUnicode_AsUTF8(co->co_filename));
+        _PyCode_SetExtra((PyObject *)co, extra_code_index,
+                         (void *)new_trampoline);
+        f = new_trampoline;
+    }
+    assert(f != NULL);
+    return f(_PyEval_EvalFrameDefault, ts, frame, throw);
+}
+#endif
+
+int
+_PyPerfTrampoline_Init(int activate)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (!activate) {
+        tstate->interp->eval_frame = NULL;
+    }
+    else {
+#ifdef HAVE_PERF_TRAMPOLINE
+        tstate->interp->eval_frame = py_trampoline_evaluator;
+        if (new_code_arena() < 0) {
+            return -1;
+        }
+#endif
+    }
+    return 0;
+}
+
+int
+_PyPerfTrampoline_Fini(void)
+{
+#ifdef HAVE_PERF_TRAMPOLINE
+    free_code_arenas();
+    perf_map_close(perf_map_file);
+#endif
+    return 0;
+}
+
+PyStatus
+_PyPerfTrampoline_AfterFork_Child(void)
+{
+#ifdef HAVE_PERF_TRAMPOLINE
+    // close file in child.
+    perf_map_close(perf_map_file);
```

Comment:
`perf_map_file` needs to be reset to NULL so that this child can save its own profiling maps via `perf_map_get_file()`